### ISSUE TITLE
String#squeeze and String#delete are slow

### ISSUE THREAD

**Issue by clayton-shopify** [mruby/mruby#4130]
The new `String#squeeze` and `String#squeeze!` methods have an execution time of O(n^2) in the worst case, because they copy the remainder of the string each time a duplicate character is encountered:
```
$ time bin/mruby -e "('a'*300000).squeeze"

real	0m14.960s
user	0m14.899s
sys	0m0.036s
```
 CRuby uses an O(n) algorithm:
```
$ time ruby -e "('a'*300000).squeeze"

real	0m0.084s
user	0m0.067s
sys	0m0.014s
```

**Comment by clayton-shopify** [comment]
`String#delete` and `String#delete!` are also slow.

mruby:
```
$ time bin/mruby -e "('a'*300000).delete('a')"

real	0m14.943s
user	0m14.866s
sys	0m0.046s
```
CRuby:
```
$ time ruby -e "('a'*300000).delete('a')"

real	0m0.113s
user	0m0.070s
sys	0m0.020s
```

**Comment by take-cheeze** [comment]
Seems calling `tr_parse_pattern` every time allocates memory every time.
Simple pattern shouldn't allocate memory.

**Comment by matz** [comment]
The real reason is calling `memove` repeatedly. It's fixed by 9e3cbaa.
If you want to try further improvement, we welcome.
